src/pybind/mgr/dashboard/models_proposal.md

```diff
@@ -0,0 +1,608 @@
+    ganesha_conf = GaneshaConf.instance(cluster_id)
+    # ...
+    return {
+      "export_id": 1,
+      #...
+    }
+```
+
+As you can see from above the two first parameters of the `set` method correspond to the path
+parameters declared in the class attribute `RESOURCE_ID`. These parameters must appear before the
+parameters used for the body data in the endpoint method declaration.
+
+With the new path models we can rewrite the above example as the following:
+
+```python
+class ExportKeyModel(PathModel):
+  cluster_id = Attr.String(
+    description="Cluster identifier")
+
+   export_id = Attr.Int(
+    description="Export ID",
+    validator=Val.Gt(0))
+
+@ApiController('/nfs-ganesha/export', Scope.NFS_GANESHA)
+class NFSGaneshaExports(RESTController):
+  # ...
+
+  @NfsTask('edit', {'cluster_id': '{key.cluster_id}', 'export_id': '{key.export_id}', 2.0)
+  def set(self, key: ExportKeyModel, export: ExportModel) -> ExportModel:
+    # export_id = int(key.export_id)  no need to do this, key.export_id is already an interger
+    ganesha_conf = GaneshaConf.instance(key.cluster_id)
+    # ...
+    # build the result model
+    return result_model
+```
+
+Since now the parameter declaration ordering does not matter we could also declare the `set` method
+as:
+
+```python
+# ...
+def set(self, export: ExportModel, key: ExportKeyModel) -> ExportModel:
+# ...
+```
+
+One important note in the above example is that we removed the `RESOURCE_ID` declaration because
+we now use the structure of the `ExportKeyModel` class to understand how to parse the path parameter
+values from the request URL. This also means that the declaration order of the attributes in the
+path model class matters. In this case `cluster_id` comes before `export_id` and therefore, if
+the request URL is of the form `https://dashboard.dom/api/nfs-ganesha/export/hello/1` then
+`cluster_id` will have the `hello` value and `export_id` will have the `1` value.
+
+
+### Query parameters
+
+Query parameters are parsed from the request URL and are encoded as python method parameters.
+
+Usually these method parameters are declared as optional parameters in the endpoints methods, as
+most of the times query parameters are not mandatory.
+
+To encode query parameters as a model, we just need to create a subclass of the `QueryModel` class.
+
+I will not add an example in this section as the class structure for the query model, and its usage
+in a controller is similar to the examples in previous sections.
+
+
+### Header parameters
+
+An HTTP request also transports data in the form of HTTP headers. Currently our controller
+infrastructure does not support the retrieval of this kind of information directly.
+To access the headers we still need to go through the cherrypy library, which breaks the controller
+infrastructure abstraction.
+
+To fix this problem we can get the headers data the same way as we get other types of data of an
+HTTP request, using a model class.
```

Comment:
Currently there is no controller in the dashboard code that uses the headers.
One reason I add the `HeaderModel` to the proposal was because it fitted well in the abstraction of representing any kind of data as a "model". Another reason is that by completely abstracting the web server technology we use below our "controller infrastrcuture" we could in theory change from using cherrypy to another solution without changing most of the code.

Regarding the way we will implement pagination is out of the scope of this proposal.

Reply:
So that means we are open to other web frameworks? I though we didn't want to move from Cherrypy.

IMHO pagination should be part of this proposal as it's a hot topic.